### ISSUE TITLE
Resource Adoption

### ISSUE THREAD

**Issue by swgillespie** [pulumi/pulumi#1662]
While pondering https://github.com/pulumi/pulumi/pull/1654, we've realized that there is an interesting operation lurking that we have tentatively called "adoption". where a user expresses in their Pulumi program that they are "adopting" a resource that was previously *not* managed by Pulumi into their program so that, in the future, it *is* managed by Pulumi.

It is an interesting and useful scenario to be able to "adopt" a resource without having to delete it. This can't be expressed in Pulumi today, because:

1. The user needs to express that a resource should be created and request that it have a specific ID - the same ID of the existing resource that a user intends to adopt.
2. The engine, when it sees an adoption where the requested ID is equivalent to the read ID, needs to not delete the resource.

This implies some changes to the language host and language projection layers of Pulumi. We should think about whether we want these changes and what they would look like.

**Comment by swgillespie** [comment]
cc @lukehoban @pgavlin @joeduffy

**Comment by lukehoban** [comment]
Going to tackle this in 0.18.

**Comment by lukehoban** [comment]
@pgavlin Could you summarize here (or in docs somewhere) the current process that can be used to accomplish this, and the plans for how we expect we'll be able to make this easier in the (near) future?

**Comment by lukehoban** [comment]
From pulumi/pulumi-kubernetes#261, note that there are scenarios where we want to adopt a resource created implicitly by another step in the deployment.  In that case, the id of the resource being adopted is not known until the first resource is deployed.  In theory this could work in the same approach sketched above, but it's a use case to make sure we think about when we design a solution here.

**Comment by lukehoban** [comment]
Resource adoption (#1662) and import (#1635) are important and we should continue to drive toward a design, but let's move these to M21 to give time to flesh out a thorough design.

**Comment by donaldguy** [comment]
Is there a hacky way to do this currently? e.g. hand editing (or jq transforming) a local state file?

**Comment by gdhagger** [comment]
@donaldguy My current hacky way of doing this is to run a `pulumi stack export > stack.json`, copy a similar resource and amend it's name everywhere, then reimport with a `pulumi stack import < stack.json`

only done it for simple resources like ip addresses mind you

**Comment by kurlzor** [comment]
> @donaldguy My current hacky way of doing this is to run a `pulumi stack export > stack.json`, copy a similar resource and amend it's name everywhere, then reimport with a `pulumi stack import < stack.json`

I tried doing this with a GKE cluster without success... (but maybe I just suck 😅)

**Comment by farnoy** [comment]
I have a workaround until this feature is shipped, example using `gcp.sql.DatabaseInstance`:

1. Use get functions to read the existing resource and bring its parameters into the stack, like this:

        export const database = gcp.sql.DatabaseInstance.get('test-database', 'name-in-gcp');
2. Run `$ pulumi up` to read the attributes and save them into the stack
3. Exchange your definition of that resource:

       export const database = new gcp.sql.DatabaseInstance('test-database', {
          name: 'name-in-gcp',
          databaseVersion: 'POSTGRES_11',
          settings: {
            tier: 'db-f1-micro'
          }
        });
    I think it's important for the exported variable in code to have the same name before & after
4. Edit your stack.json file to remove the key-paid `"external": true` from it. You can either modify the file locally, or do `stack export` followed by `stack import`
5. Run `$ pulumi refresh`, the parameters of this resource will be slightly different in the stack.json file. Pulumi will output something like: "Resources: 2 unchanged", but the attributes are modified under the hood.
6. That's it, this resource is now properly managed by Pulumi. To test it, I changed the tier to `db-g1-small` and ran `$ pulumi up` and it is being properly updated. The first time I made a change to configuration of my database, it dropped some other fields that I was not managing explicitly, specific to this resource, like "crashSafeReplication". But every following modification should have a small diff with just the parameters you change.

**Comment by lukehoban** [comment]
Thanks @farnoy for summarizing this!  Indeed, I've worked with a few users to manually do something similar.  We're aiming to improve this experience, but it is great to have the currently available approach documented like this.

**Comment by agolomoodysaada** [comment]
This would be especially useful when migrating from Helm to pulumi. I imagine an API like

`pulumi adopt tiller` that talks to the existing Tiller server, pulls all existing Helm releases and generates a file for each release and their corresponding values. The output pulumi files can then be easily modified and added to CI.

**Comment by gambrose** [comment]
The workaround did not work for me when trying to adopt an `azure.core.ResourceGroup`

I got the resource group `azure.core.ResourceGroup.get('rg', pulumi.output(azure.core.getResourceGroup({ name: 'enmodus-beta' })).id);`
ran `pulumi up` changed the definition to use `new azure.core.ResourceGroup` and removed the `external": true` by export, edit, import method.
Running `pulumi up` works with no changes but when I edit the resource group tags and try and apply the change I get:
```
azure:core:ResourceGroup (rg):
    error: Plan apply failed: 1 error occurred:
        * updating urn:pulumi:enmodus-beta::wattzo::azure:core/resourceGroup:ResourceGroup::rg: A resource with the ID "/subscriptions/ee32d975-a431-4b20-a603-0fb9071dad0f/resourceGroups/enmodus-beta" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_group" for more information.
```

**Comment by clstokes** [comment]
For anyone needing to do this before this gets implemented, I took @farnoy's great writeup and expanded it a bit for another user of ours - https://gist.github.com/clstokes/62d8ff84ada8cd414a88d4f1030f5ddc.
